pmtiles/server_metrics.go

```diff
@@ -0,0 +1,203 @@
+	} else {
+		buildTimeMetric.Set(0)
+	}
+}
+
+type metrics struct {
+	// overall requests: # requests, request duration, response size by archive/status code
+	requests        *prometheus.CounterVec
+	responseSize    *prometheus.HistogramVec
+	requestDuration *prometheus.HistogramVec
+	// dir cache: # requests, hits, cache entries, cache bytes, cache bytes limit
+	dirCacheEntries    prometheus.Gauge
+	dirCacheSizeBytes  prometheus.Gauge
+	dirCacheLimitBytes prometheus.Gauge
+	dirCacheRequests   *prometheus.CounterVec
+	// requests to bucket: # total, response duration by archive/status code
+	bucketRequests        *prometheus.CounterVec
+	bucketRequestDuration *prometheus.HistogramVec
+	// misc
+	reloads *prometheus.CounterVec
+}
+
+// utility to time an overall tile request
+type requestTracker struct {
+	start   time.Time
+	metrics *metrics
+}
+
+func (m *metrics) startRequest() *requestTracker {
+	return &requestTracker{start: time.Now(), metrics: m}
+}
+
+func (r *requestTracker) finish(archive, handler string, status, responseSize int, logDetails bool) {
+	labels := []string{archive, handler, strconv.Itoa(status)}
+	r.metrics.requests.WithLabelValues(labels...).Inc()
+	if logDetails {
+		r.metrics.responseSize.WithLabelValues(labels...).Observe(float64(responseSize))
+		r.metrics.requestDuration.WithLabelValues(labels...).Observe(time.Since(r.start).Seconds())
+	}
+}
+
+// utility to time an individual request to the underlying bucket
+type bucketRequestTracker struct {
+	start   time.Time
+	metrics *metrics
+	archive string
+}
+
+func (m *metrics) startBucketRequest(archive string) *bucketRequestTracker {
+	return &bucketRequestTracker{start: time.Now(), metrics: m, archive: archive}
+}
+
+func (r *bucketRequestTracker) finish(status string) {
+	labels := []string{r.archive, status}
+	r.metrics.bucketRequests.WithLabelValues(labels...).Inc()
+	r.metrics.bucketRequestDuration.WithLabelValues(labels...).Observe(time.Since(r.start).Seconds())
+}
+
+// misc helpers
+
+func (m *metrics) reloadFile(name string) {
+	m.reloads.WithLabelValues(name).Inc()
+}
+
+func (m *metrics) initCacheStats(limitBytes int) {
+	m.dirCacheLimitBytes.Set(float64(limitBytes))
+	m.updateCacheStats(0, 0)
+}
+
+func (m *metrics) updateCacheStats(sizeBytes, entries int) {
+	m.dirCacheEntries.Set(float64(entries))
+	m.dirCacheSizeBytes.Set(float64(sizeBytes))
+}
+
+func (m *metrics) cacheRequest(archive, status string) {
+	m.dirCacheRequests.WithLabelValues(archive, status).Inc()
+}
+
+func register[K prometheus.Collector](logger *log.Logger, metric K) K {
+	if err := prometheus.Register(metric); err != nil {
+		logger.Println(err)
+	}
+	return metric
+}
+
+func createMetrics(scope string, logger *log.Logger) *metrics {
+	namespace := "pmtiles"
+	durationBuckets := prometheus.DefBuckets
+	kib := 1024.0
+	mib := kib * kib
+	sizeBuckets := []float64{1.0 * kib, 5.0 * kib, 10.0 * kib, 50.0 * kib, 100 * kib, 200 * kib, 500 * kib, 1.0 * mib, 2.5 * mib, 5.0 * mib}
```

Comment:
Caddy's default for response sizes are:

>  256b, 1kiB, 4kiB, 16kiB, 64kiB, 256kiB, 1MiB, and 4MiB.

but I figure we might want more resolution around the couple-hundred KB range for tiles

Reply:
agreed, if we had to hardcode I would get even more resolution around 100-500kb range since Tippecanoe defaults to that as a tile size limit.

Reply:
👍  changing to: 

- 1kb
- 5kb
- 10kb
- 25kb
- 50kb
- 100kb
- 250kb
- 500kb
- 1mb